client/app/visualizations/chart/Renderer/PlotlyChart.jsx

```diff
@@ -0,0 +1,51 @@
+import { isArray, isObject } from 'lodash';
+import React, { useState, useEffect } from 'react';
+import { RendererPropTypes } from '@/visualizations';
+import resizeObserver from '@/services/resizeObserver';
+
+import getChartData from '../getChartData';
+import { Plotly, prepareData, prepareLayout, updateData, applyLayoutFixes } from '../plotly';
+
+export default function PlotlyChart({ options, data }) {
+  const [container, setContainer] = useState(null);
+
+  useEffect(() => {
+    if (container) {
+      const plotlyOptions = { showLink: false, displaylogo: false };
+
+      const chartData = getChartData(data.rows, options);
+      const plotlyData = prepareData(chartData, options);
+      const plotlyLayout = prepareLayout(container, options, plotlyData);
+
+      // It will auto-purge previous graph
+      Plotly.newPlot(container, plotlyData, plotlyLayout, plotlyOptions).then(() => {
+        applyLayoutFixes(container, plotlyLayout, (e, u) => Plotly.relayout(e, u));
+      });
+
+      container.on('plotly_restyle', (updates) => {
```

Comment:
Shouldn't this be moved to the below `useEffect()`?

Reply:
I'm not sure I understand your point. It uses some variables from the same scope, so cannot be moved to another `useEffect` without moving that variables to upper scope. If you mean that both `useEffect` calls should be merged - probably yes, but current variant is a sort of optimization: first `useEffect` will re-build chart when container, options and/or data changes - `Plotly.newPlot` will purge old chart if needed. And we need to call `Plotly.purge` only when container changes.

Reply:
> It uses some variables from the same scope

Perhaps one of us is misunderstanding useEffect's 2nd argument usage.
AFAIK, it's used as "componentshouldupdate". You can use any variables in the useEffect scope without adding them to the 2nd argument. Am I wrong about this?

My concern here is that `container.on` is being re-declared needlessly every time `options` and `data` change, but perhaps needs to only when `container` does.

Reply:
`container.on` should be called after each `Plotly.newPlot` because Plotly does a sort of shitty thing: it extends DOM element which is a plot container and adds new properties and methods to it, like `on`. So when you call `Plotly.newPlot` - it will destroy previous plot and unbind all events, and then you have to re-bind everything again. If that was a regular DOM events (`addEventListener`) - then yes, I totally agree with you